debug/org.eclipse.debug.core/core/org/eclipse/debug/core/ILaunchAttribute.java

```diff
@@ -0,0 +1,170 @@
+ * <li>{@link ILaunchConfigurationWorkingCopy#setAttribute(String, String)} and
+ * similar operations</li>
+ * <li>Connecting {@link ILaunchConfiguration} attributes with preferences</li>
+ * <li>Representing {@link ILaunchConfiguration} attributes in UI</li>
+ * </ul>
+ *
+ * @since 3.23
+ */
+public interface ILaunchAttribute<V> {
+
+	/**
+	 *
+	 * @return identity for defined attribute
+	 */
+	ILaunchAttributeIdentity identity();
+
+	/**
+	 * Reads the value of launch configuration attribute and tries to substitute
+	 * variables for String values. <br/>
+	 * Returns the attribute value or fails with {@link CoreException}, may
+	 * return <code>null</code> value
+	 *
+	 * @param configuration the launch configuration to read attribute from,
+	 *            must not be <code>null</code>
+	 * @return the attribute value, may return <code>null</code>
+	 */
+	V read(ILaunchConfiguration configuration) throws CoreException;
+
+	/**
+	 * "Probes" the value of launch configuration attribute. <br/>
+	 * Returns {@link Optional} in case of read failure or <code>null</code>
+	 * value to let caller decide how to proceed
+	 *
+	 * @param configuration the launch configuration to read attribute from,
+	 *            must not be <code>null</code>
+	 * @return the {@link Optional} with attribute value
+	 */
+	Optional<V> probe(ILaunchConfiguration configuration);
+
+	/**
+	 * Writes the supplied value to the launch configuration working copy
+	 *
+	 * @param working the launch configuration working copy to write value to,
+	 *            must not be <code>null</code>
+	 * @param value the value, may be <code>null</code>
+	 */
+	void write(ILaunchConfigurationWorkingCopy working, V value);
+
+	/**
+	 * creates an instance of {@link ILaunchAttribute} using primitives (useful
+	 * for existing code)
+	 *
+	 * @param <V> the value type for the launch attribute
+	 * @param identity the id of the attribute, must not be <code>null</code>
+	 * @param type the value type of the attribute, must not be
+	 *            <code>null</code>
+	 * @param value the {@link Function} to calculate of default value of the
+	 *            attribute from {@link ILaunchConfiguration}, must not be
+	 *            <code>null</code>
+	 * @param name the name of the attribute, must not be <code>null</code>
+	 * @return created instance of {@link ILaunchAttribute}
+	 */
+	static <V> ILaunchAttribute<V> of(ILaunchAttributeIdentity identity, Class<V> type, Function<ILaunchConfiguration, V> value, String name) {
+		return new ConsiderStringVariables<>(identity, type, value, name, name);
+	}
+
+	/**
+	 * Default implementation for {@link ILaunchAttribute} that tries to
+	 * substitute variables for launch attributes with {@link String} type
+	 *
+	 * @see IStringVariableManager
+	 */
+	record ConsiderStringVariables<V>(ILaunchAttributeIdentity identity, Class<V> type, Function<ILaunchConfiguration, V> defaults, String name, String description) implements ILaunchAttribute<V> {
```

Comment:
Records should really be used with great caution in APIs, because as far as I understand it, even just adding a component could be a binary incompatible change when considering newer lange-features like pattern matching.

So I would just make this a local class/record of the static `of` factory

Reply:
extracted to package local type